<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
  <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
  <meta http-equiv="Content-Style-Type" content="text/css" />
  <meta name="generator" content="pandoc" />
  <title>Glossary</title>
  <style type="text/css">code{white-space: pre;}</style>
  <link rel="stylesheet" href="manual.css" type="text/css" />
</head>
<!-- BODY BEGIN ------------------------------>
<body>
<div class="toc contentbox">
    <span class="tocheader">Table of contents</span>
<ul>
<li><a href="index.html">Package overview</a></li>
<li><a href="00b_shell-prerequisites.html">Dependencies of the scripts</a>
<li><a href="00c_input-files.html">Input files</a>
<li><a href="01_genome-preparation.html">Genome Preparation</a>
<li><a href="02_read-generation.html">Generation of simulated read sets</a>
<li><a href="03_read-mutation.html">Mutation of reads</a>
<li><a href="04_mapping.html">Generate mapper calls</a>
<li><a href="05_sam-parsing.html">Parsing of SAM files</a>
<li><a href="06_parameter_influence.html">Comparing multiple mapper runs</a>
<li><a href="A01_workflow.html">Appendix A1 -- Tutorial script</a>
<li><a href="A02_advanced_read-generation.html">Appendix A2 -- Advanced read generation</a>
<li><a href="A03_scripts.html">Appendix A3 -- Overview of the scripts by category</a>
<li class="tocthis"><a href="B99_glossary.html">Glossary</a>
</ul>
</div>
<div class="content contentbox">
<div id="header">
<h1 class="title">Glossary</h1>
</div>
<!-- Local Table of contents -->
<!--Back to table of contents -->
<!-- CONTENT -------------------------------------->
<p>Some terms are explained here, which are used in the rest of the manual and may have a special meaning:</p>
<p>~table: A text table is the most widely used data format in this project. It is a simple text file, where the columns are separated by a special character. Usually this is a tabulator (tab) character, but sometimes space-separated tables are also used.</p>
<p>~FASTA record: A FASTA record is a block in a fasta file which is delimited by two &gt; signs. It consists of the following parts:</p>
<ol class="incremental" style="list-style-type: decimal">
<li>The header line: &gt;.................(line end).
<ol class="incremental" style="list-style-type: lower-alpha">
<li>The FASTA identifier, which is the first 'word' of the header line. E.g. in a header line</li>
</ol>
<blockquote>
<p>A1 Felis catus genome ID:012345...</p>
</blockquote>
<pre><code>the FASTA identifier is &#39;A1&#39;.</code></pre>
<ol class="incremental" start="2" style="list-style-type: lower-alpha">
<li>The FASTA record description: All the text between identifier and line end.</li>
</ol></li>
<li>Biological sequence data, IUPAC one-character-code. Should not be wider than 80 characters and may contain whitespace. E.g. nucleotides: ACCTCTCTACCT...</li>
</ol>
<p>~FASTA identifier: -&gt; FASTA record</p>
<p>~FASTA description: -&gt; FASTA record</p>
<p>~file offset: This is the distance in bytes from the beginning of a file. The first character of a file has the offset 0.</p>
<p>~offset: -&gt; file offset</p>
<p>~i-based index: An index is i-based if it starts counting with number i. For example, the mapping position (pos) field of a SAM file is a 1-based index. If the position 1 is written there, the read maps to the first base of the genome. Conversely, if the index were 0-based, the first base of the genome would be referenced with the number 0.</p>
<p>~0-based: -&gt; i-based index</p>
<p>~1-based: -&gt; i-based index</p>
<p>~standard input: Many commands expect input on this stream. Input can be provided either by typing into the console, by using the &lt; operator to provide input from a file or by the | operator, which forwards the content on standard output of a previous command to standard input of this command. Refer to the &quot;Redirection&quot; section of your shell for more information.</p>
<p>~standard output: All output a command writes is by default redirected to the standard output or standard error streams. Per convention, standard output is used for the results of the program, whereas standard error is used for status and error messages. Content on standard output can be written in a file by using the &gt; operator or redirected to standard input of another command by using the | operator. Standard error output can be redirected to a file using the 2&gt; operator. Consult the &quot;Redirection&quot; section of your shell for more information.</p>
<p>~standard error: -&gt; standard output</p>
<p>~command line argument: a value which is written on the command line behind the name of the program which shall be invoked. They are forwarded to the program and influence it. The valid parameters of a program are described in the program's manual. (often accessible via the <code>--help</code> command line argument.</p>
<p>~command line switch: -&gt; command line argument</p>
</div>
<div class="contentbox linktotop"><a href="#top">Go to top</a></div> 
</body>
</html>
